### PR TITLE
Add FullFileName and NumCaptured records

### PR DESCRIPTION
**Pull request by jsouter**
Some changes to help with the Ophyd's HDF writing device.
Adds FullFileName record as in AreaDetector and NumCaptured record.
Also fix logic issue in the "async for data in self._client.data" loop, where an exception is raised when an EndData packet is received.

### CODE REVIEW

**Comment by jsouter** [comment]
I should probably have a look at the tests too.

**Comment by jsouter** [comment]
Added a few lines in to test the new records. Unsure of how best to handle NumCaptured, it may be better if it can only get incremented once we've confirmed that the frame has been successfully written into the HDF5 file, rather than incrementing it when the frame is obtained from the stream.

**Review by evalott100 — APPROVED** [review]
Just a few small things. Have you tested this on a real panda - worth doing before we merge.

**Comment by jsouter** [comment]
> Just a few small things. Have you tested this on a real panda - worth doing before we merge.

Yep, only with the one layout though

File: src/pandablocks_ioc/_hdf_ioc.py
Comment (evalott100) [reply src/pandablocks_ioc/_hdf_ioc.py:92]: Hmmm if that's so maybe we should check with @coretl

Comment (evalott100) [reply src/pandablocks_ioc/_hdf_ioc.py:92]: Talked to Tom and we agreed this would be nice, but we don't want it to be too different from Odin. What does Odin currently have?

Comment (jsouter) [reply src/pandablocks_ioc/_hdf_ioc.py:92]: Pretty sure Odin inherits that bit from ADCore so FileName(_RBV), FIlePath(_RBV) and FullFileName(_RBV)

Comment (evalott100) [reply src/pandablocks_ioc/_hdf_ioc.py:92]: Ah yea this is pretty baked in, would be a real pain to change now. I'll defer to Tom as to whether it's worth changing.

``` python
# in ophyd plugins.py
class FileBase(Device):
    ...
    file_name = Cpt(SignalWithRBV, "FileName", string=True, kind="config")
    file_number = Cpt(SignalWithRBV, "FileNumber")
    file_path = Cpt(
        EpicsPathSignal, "FilePath", string=True, kind="config", path_semantics="posix"
    )
    full_file_name = Cpt(EpicsSignalRO, "FullFileName_RBV", string=True, kind="config")
```

Comment (evalott100) [reply src/pandablocks_ioc/_hdf_ioc.py:92]: Another thing @coretl brought up is that it'd be good to include a `file_number` pv in this PR (in the same way it's used in areadetector).

Comment (coretl) [reply src/pandablocks_ioc/_hdf_ioc.py:92]: Ok, let's stick with the ad names, but don't add a format or a number, just FileName(_RBV), FIlePath(_RBV) and FullFileName_RBV

File: src/pandablocks_ioc/_hdf_ioc.py
Comment (jsouter) [reply src/pandablocks_ioc/_hdf_ioc.py:53]: That's fair, I guess I was thinking it was needed inside the finally block but it should still be accessible anyway, will undo this change in the next push.

**Comment by evalott100** [comment]
@jsouter I've pulled your changes to a new branch that will introduce all the other changes we discussed. If you'll be working on ophyd stuff are we good to close this?

**Comment by jsouter** [comment]
I believe so, yeah. I'll keep an eye on your branch and reflect whatever changes to PV names there are in my ophyd-async branch.

**Pull request closed without merging**
